### ISSUE TITLE
dockerTools.buildLayeredImage can fail when other nix builds are in progress

### ISSUE THREAD

**Issue by purefn** [NixOS/nixpkgs#75888]
**Describe the bug**
In [pkgs/build-support/docker/store-path-to-layer.sh](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/store-path-to-layer.sh), `tar` is run against `/nix/store`. But if you have multiple packages being built in parallel, and something is added to `/nix/store`, tar will fail with a `tar: /nix/store: file changed as we read it` message.

**To Reproduce**
Build a derivation that uses `buildLayeredImage` while at the same time building other derivations.

**Expected behavior**
No errors to occur.

**Additional context**
This is a similar problem to the one described #34137

**Comment by purefn** [comment]
@nlewo, do you have any ideas about how to fix this? I'm wondering if maybe creating a mirrored, hard link, directory tree (similar to what `lndir` does but with hard links) in a temporary directory, and then running tar against that tree would address this issue. Or is there a better way?

**Comment by purefn** [comment]
Alternatively, we could pass `--warning=no-file-changed`, wrap the `tar` usage in `set +x`/`set -x` and manually check the exit code of running `tar`. Only if the exit code is `0` or `1` do we say it succeeded, otherwise, if it is `2`, it failed and the script fails.

**Comment by nlewo** [comment]
I don't think we should read `/nix/store` during the build of a derivation since they are not immutable. So, i would prefer a solution that doesn't rely on these directories.
The https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/store-path-to-layer.sh#L12 tar command is only used to create a `/nix/` and a `/nix/store` directory in the tar archive. I think it would be possible to create these directories in the build directory, and tar them by transforming the path to move them to `/` in the tar file.
I would suggest something such as:
```
$ mkdir -p nix/store 
$ tar  --transform='s,nix,/nix,' -cf  test.tar nix
$ tar -tf test.tar 
tar: Removing leading `/' from member names
/nix/
/nix/store/
``` 
Would this fix your issue?

**Comment by purefn** [comment]
Not completely because that's only one use of `tar` that could fail. The second can fail in the same way. I experimented with it and was able to make `tar` produce the same error.

It can also fail in building the customization later as mentioned here https://github.com/NixOS/nixpkgs/pull/75911#issuecomment-567723248

**Comment by purefn** [comment]
For the second `tar` command, I think if we can not need to reference `/nix/store` that will get us what we need. 

I've been experimenting with this combination of tar commands

```
mkdir -p nix/store                                                                                                                                             
tar -cf "$layerPath/layer.tar"  \                                                                                                                              
    --mtime="@$SOURCE_DATE_EPOCH" \                                                                                                                            
    --owner=0 --group=0 \                                                                                                                                      
    --transform='s,nix,/nix,' \                                                                                                                                
    nix                                                                                                                                                        
                                                                                                                                                               
n=$(basename "$@")                                                                                                                                       
tar -C /nix/store -rpf "$layerPath/layer.tar" \                                                                                                                
    --hard-dereference --sort=name \                                                                                                                           
    --mtime="@$SOURCE_DATE_EPOCH" \                                                                                                                            
    --owner=0 --group=0 \                                                                                                                                      
    --transform="s,$n,/nix/store/$n," \                                                                                                                        
    "$n"                                                                                                                                                       
```

The first is as you suggested. The second does the packing from `/nix/store`. We use `basename` to get the subpath, and use the transform to prepend `/nix/store/` to each of the paths in the archive. This way `tar` isn't trying to monitor everything in `/nix/store` and will only care about our specific path.

I'll update my PR with that approach.

**Comment by purefn** [comment]
I've been testing the PR quite a bit and it does fix the issue with the script very nicely.

However, it does't completely fix the problem with building layered images in parallel because the building the final tarball at https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/docker/default.nix#L628 can fail in a similar way.

```
docker-image-learning-server-nix.tar.gz> Cooking the image...
docker-image-learning-server-nix.tar.gz>  tar: ./6773e5a7a063b5386300f1675b617904d21ee4b8c7f9d75b275c67a62cf18e0a/layer.tar: file changed as we read it
```

I'm absolutely baffled by what `tar` might be seeing. Those files are built as part of a prior derivation so they should not be changed at all. I've compared the file attributes and checksum of the `layer.tar` files from before and after that `tar` command is run and confirmed that nothing actually changes with the file(s) in question. I have no idea why building multiple layered images would even cause this issue, as those derivations aren't sharing the granular layers derivation in any way.

At this point I have to move on and will just be reverting to building our layered images separately from the derivations that compose them and using `--max-jobs 1` when building the layered docker images. :cry:

**Comment by nlewo** [comment]
Hm, I don't understand your issue:( So, some raw ideas...
- Did you compare the links (sym vs hard)? 
- Do you have the Nix option `auto-optimise-store` set (not sure it's relevant)?
- Could you try without it (not sure it's relevant)?
- Is tar able to tell you what changed (i don't think so)?

**Comment by nlewo** [comment]
Are you able to reproduce with `buildImage`?

**Comment by purefn** [comment]
> * Do you have the Nix option `auto-optimise-store` set (not sure it's relevant)?
> * Could you try without it (not sure it's relevant)?

Yes, I do! I just tested with it turned off and was unable to reproduce the issue after several test runs, where before I couldn't even get it to succeed once. So I'm pretty confident that is the source of the problem.

This means it won't be an issue on our build machines as I don't think we'll have that enabled there, but it will be an issue for any users trying to build docker images locally with `auto-optimise-store` turned on - like me. I'm hesitant to tell people to turn it off, or turn it off myself, because it does save me significant disk space savings.

This brings me back to my original approach of having tar and the script ignore these sorts of errors. We *know* that these files are in the nix store and are therefore immutable, other than having more hard links added. That means it should be safe to ignore such errors. In general I don't like ignoring errors, but in this case we can reasonably say there is only one way this error would present itself and we don't really care about that failure case. Is there another way we can make this work for users with this option enabled?

**Comment by nlewo** [comment]
Ok, cool. We know have a better idea on what's going on!

Like you, I don't like to ignore these errors. I also think we can miss other errors. So, let's try another way.

I'm wondering what is the tar behavior when we submit the list of files to tar instead. The idea would be to do something such as:
```
find /nix/store/PATH > list.txt
tar -T list.txt -c...f layer
```

Maybe tar will not detect the file changes since it didn't generate the list of files. Could you try this way?

**Comment by purefn** [comment]
That has the same result. `tar` seems to think the number of hard links is somehow important, no matter what. I thought maybe removing `--hard-dereference` might help, but that doesn't help either. 

This appears to have come up in other contexts, and the [solution](https://patchwork.openembedded.org/patch/91015/) there was to ignore the `tar` exit code. I'm afraid there is little else we can do.

**Comment by nlewo** [comment]
Ok... So, let's go with your solution then!

**Comment by nlewo** [comment]
IIUC, we then have two issues here:
1. tar fails because `/nix/store` is modified during tar process. This is because files can be added in the /nix/store directory during the tar process.
2. tar fails because some files are transformed to hardlinks during the tar process. This happens when the Nix option `auto-optimize-option` is `true`.

I think #75911 addresses the issue 1 without having to do the hack with the tar return code.

So, I think it could be nice to open another PR to address 2. with your initial solution (by checking the tar return code).

I don't think we can write regression tests for these issues because they are related the Nix environment itself, which we are not controlling in our test frameworks... This is also why I think it's nice to distinguish these two issues in our code base.

**Comment by flokli** [comment]
https://github.com/NixOS/nixpkgs/pull/91084 has been merged, which might fix this issue. Can you have a look?

**Comment by wwilson** [comment]
We ran into this issue at Antithesis, and I think have some insight into the core problem.

First of all, it doesn't only affect dockerTools.buildLayeredImage. In principle, any Nix build that uses GNU Tar can have this happen. The core problem is that tar considers an updated ctime to be a "modification" to an input file, however all kinds of totally nondeterministic and unpredictable things can cause a change to the ctime: for instance another process adding or removing a hard link to the file in question. This is why auto-optimise-store can trigger the issue, but the thing is all kinds of other stuff running on your system could do it! It's Tar's behavior here that is unreasonable.

Tar itself seems to be somewhat aware that this is an overly conservative check. It uses exit code 1 to signal "warnings", and then proceeds to go ahead with the archive operation. Throughout the tar code base, it's 2 that is used to signal actual fatal errors. On the other hand, we don't want to patch Dockertools etc. to ignore tar exit code 1, since there could be other tar "warnings" that are sound and that should trigger Nix builds to fail.

For now we've done the following as an absolutely minimal patch to tar, only changing the exit code in this ctime modification case:

```diff
diff -u --recursive tar-1.34/src/create.c tar-1.34-patched/src/create.c
--- tar-1.34/src/create.c	2021-02-04 09:00:33.000000000 -0500
+++ tar-1.34-patched/src/create.c	2021-10-28 13:43:10.651890357 -0400
@@ -1836,7 +1836,8 @@
 	      WARNOPT (WARN_FILE_CHANGED,
 		       (0, 0, _("%s: file changed as we read it"),
 			quotearg_colon (p)));
-	      set_exit_status (TAREXIT_DIFFERS);
+	      //Antithesis: ctime changes for harmless reasons like hardlink creation, breaking nix builds nondeterministically
+	      //set_exit_status (TAREXIT_DIFFERS);
 	    }
 	  else if (atime_preserve_option == replace_atime_preserve
 		   && fd && (is_dir || original_size != 0)
```

The trouble is that pretty much everything in nixpkgs depends on tar, so we don't want to override pkgs.gnutar in an overlay, because doing so would mean we lose all benefit from Hydra. It's really up to the Nix community to decide how it wants to approach this. However my quick search of github suggests that this is actually the underlying cause of many different open issues related to build flakiness, some involving dockertools, some not.

**Comment by nlewo** [comment]
Excepting if the `auto-optimize` feature is enabled, I don't know what could update a output store path. Do you have another example?

Regarding your patch, we could apply it on `dockerTools` only. Since it doesn't affect the output hash of produced images, we could revert it easily if there is a better solution. (Applying it globally is a much more complex decision :confused:)

**Comment by wwilson** [comment]
A case unrelated to auto-optimize-store that we ran into recently: we built a container using dockertools that contained some of the same PAM definition files as the host. All kinds of asynchronous system processes add and remove hardlinks to the PAM files in the Nix store, and it's easy to cause a dockertools build to fail this way.

Applying the patch specifically to the copy of tar used by dockerTools is exactly what we've done in house, but that's only because we don't want to rebuild all of nixpkgs. In fact I think the Nix project _should_ patch pkgs.gnutar globally, because in theory this same issue could cause any build anywhere in Nix that happens to use Tar to fail nondeterministically.

**Comment by wwilson** [comment]
What are the steps to getting this applied globally? We think this is potentially a big source of flakiness in Nix builds, and would prefer not to maintain our own separate cached build of all of nixpkgs.
